src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
     /// <inheritdoc />
     public Stream GetDecompressionStream(Stream stream)
     {
-        return new DeflateStream(stream, CompressionMode.Decompress, leaveOpen: true);
+        return new ZLibOrDeflateStream(stream);
+    }
+
+    // As described in RFC 2616, the deflate content-coding token represents the "zlib" format
+    // (RFC 1950) in combination with the "deflate" compression algorithm (RFC 1951). However,
+    // in practice, it is also possible for raw, unwrapped deflate compression to be used with
+    // "deflate" as the content-encoding. This class lets us wrap either a zlib- or deflate-
+    // compressed stream and delay figuring out what it is until the first read.
+    internal sealed class ZLibOrDeflateStream : Stream
+    {
+        private readonly PeekFirstByteReadStream _stream;
```

Comment:
how many of these ZLibOrDeflateStream are created? is it worth the effort to collapse these two fields. checking the type, and if not null then getting the inner from the BaseStream property.